migration/src/main/java/kin/sdk/migration/MigrationManager.java

```diff
@@ -1,60 +1,404 @@
+		int retryCounter = 0;
+		while (true) {
+			try {
+				Log.d(TAG, "burnAccount: ");
+				ITransactionId transactionId = account.sendBurnTransactionSync(publicAddress);
+				return transactionId.id() != null && !transactionId.id().isEmpty();
+			} catch (OperationFailedException e) {
+				Throwable cause = e.getCause();
+				// Check if it is an http exception with 5xx code and if yes then retry until max tries reached.
+				if (cause instanceof HttpResponseException) { // TODO: 30/12/2018 maybe need to catch more exception for the retry? like java.net.ConnectException?
+					HttpResponseException httpException = (HttpResponseException) cause;
+					if (httpException.getStatusCode() >= 500 && retryCounter < MAX_RETRIES) {
+						retryCounter++;
+						Log.d(TAG, "burnAccount: retry number " + retryCounter);
+						continue;
+					}
+				}
+				throw new MigrationFailedException(e);
+			}
+		}
+	}
+
+	private void migrateToNewKin(final String publicAddress, final MigrationManagerListener migrationManagerListener) {
+		Log.d(TAG, "migrateToNewKin: sending the request to migrate");
+		sendRequest(URL_MIGRATE_ACCOUNT_SERVICE + publicAddress, new Callback() {
+			@Override
+			public void onFailure(@NonNull Call call, @NonNull IOException e) {
+				fireOnError(migrationManagerListener, e);
+			}
+
+			@Override
+			public void onResponse(@NonNull Call call, @NonNull Response response) {
+				if (response.isSuccessful()) {
+					fireOnReady(migrationManagerListener, initNewKin(), true);
+				} else {
+					handleMigrationException(response.body(), migrationManagerListener);
+				}
+			}
+		});
+	}
+
+	// Handle the migration exceptions, if an exception that is not solvable then throw it, else just handle it.
+	private void handleMigrationException(ResponseBody body, MigrationManagerListener migrationManagerListener) {
+		// check if account has been migrated successfully and if yes then complete the process and update the persistent state.
+		Log.d(TAG, "generateMigrationException: ");
+		boolean handled = false;
+		MigrationFailedException exception = new MigrationFailedException(
+			"Migration not completed due to an unexpected exception");
+		if (body != null) {
+			try {
+				Type type = new TypeToken<Map<String, String>>() {
+				}.getType();
+				Map<String, String> error = new Gson().fromJson(body.string(), type);
+				final String code = error.get("code");
+				String message = error.get("message");
+				if (code != null) {
+					switch (code) {
+						case "4001":  // account not burned
+							exception = new MigrationFailedException(message);
+							break;
+						case "4002":  // account was already migrated
+							fireOnReady(migrationManagerListener, initNewKin(), true);
+							handled = true;
+							break;
+						case "4003":  // public address is not valid, meaning the format of it is not valid.
+							exception = new MigrationFailedException(message);
+							break;
+						case "4041":  // account was not found
```

Comment:
I talked with Ron about it and it is probably won't happen but if it will then currently we are just handling it just like a regular account not found, meaning there is no account so we will just give you the new kin sdk.
We can talked with Ron about it again.